custom_components/unifiprotect/entity.py

```diff
@@ -31,22 +33,35 @@ def __init__(
         if description and not hasattr(self, "entity_description"):
             self.entity_description = description
         if not hasattr(self, "device"):
-            self.device: ProtectAdoptableDeviceModel = device
+            self.device: ProtectDeviceModel = device
         self.protect: ProtectApiClient = protect
         self.protect_data: UnifiProtectData = protect_data
         if description is None:
             self._attr_unique_id = f"{self.device.id}"
         else:
             self._attr_unique_id = f"{self.device.id}_{description.key}"
-        self._attr_device_info = DeviceInfo(
-            name=self.device.name,
-            manufacturer=DEFAULT_BRAND,
-            model=self.device.type,
-            via_device=(DOMAIN, self.protect.bootstrap.nvr.mac),
-            sw_version=self.device.firmware_version,
-            connections={(dr.CONNECTION_NETWORK_MAC, self.device.mac)},
-            configuration_url=self.device.protect_url,
-        )
+
+        if isinstance(self.device, NVR):
+            self._attr_device_info = DeviceInfo(
+                connections={(dr.CONNECTION_NETWORK_MAC, self.device.mac)},
+                identifiers={(DOMAIN, self.device.mac)},
+                manufacturer=DEFAULT_BRAND,
+                name=self.device.name,
+                model=self.device.type,
+                sw_version=str(self.device.version),
+                configuration_url=self.device.api.base_url,
+            )
+        else:
+            assert isinstance(self.device, ProtectAdoptableDeviceModel)
+            self._attr_device_info = DeviceInfo(
+                name=self.device.name,
+                manufacturer=DEFAULT_BRAND,
+                model=self.device.type,
+                via_device=(DOMAIN, self.protect.bootstrap.nvr.mac),
+                sw_version=self.device.firmware_version,
+                connections={(dr.CONNECTION_NETWORK_MAC, self.device.mac)},
+                configuration_url=self.device.protect_url,
+            )
```

Comment:
Can we avoid an `if` here by using a different base class for `NVR` entities?

Reply:
That would require bubbling that up to Binary Sensors and Sensors as well though. But I was thinking about possibly look into a way to make clean up making the entities.

Reply:
Yeah, I think this is the best it can be for now. I really like how binary_sensor.py works though and would love to generalize that pattern for all of the entities to make it easier to use different flows/classes for different device types.